src/boundary.cpp

```diff
@@ -113,15 +113,34 @@ edge_tuple edge_iterate(const NumpyMatrix &distMat, const int slope,
   return edge_vec;
 }
 
-edge_tuple generate_tuples(const std::vector<int> &assignments, const int within_label) {
+edge_tuple generate_tuples(const std::vector<int> &assignments,
+                           const int within_label,
+                           bool self,
+                           const int num_ref,
+                           const int int_offset) {
     const size_t n_rows = assignments.size();
     const size_t n_samples = 0.5 * (1 + sqrt(1 + 8 * (n_rows)));
     edge_tuple edge_vec;
-    for (long row_idx = 0; row_idx < n_rows; row_idx++) {
-        if (assignments[row_idx] == within_label) {
-            long i = calc_row_idx(row_idx, n_samples);
-            long j = calc_col_idx(row_idx, i, n_samples);
-            edge_vec.push_back(std::make_tuple(i, j));
+    if (self) {
+        for (long row_idx = 0; row_idx < n_rows; row_idx++) {
+            if (assignments[row_idx] == within_label) {
+                long i = calc_row_idx(row_idx, n_samples);
+                long j = calc_col_idx(row_idx, i, n_samples) + int_offset;
+                i = i + int_offset;
+                long min_node = std::min(i,j);
+                long max_node = std::max(i,j);
+                edge_vec.push_back(std::make_tuple(min_node, max_node));
+            }
+        }
+    } else {
+        for (long row_idx = 0; row_idx < n_rows; row_idx++) {
+            if (assignments[row_idx] == within_label) {
+                unsigned long i = row_idx % num_ref + int_offset;
+                unsigned long j = static_cast<size_t>(row_idx / (float)num_ref + 0.001f) + num_ref + int_offset;
+                long min_node = std::min(i,j);
+                long max_node = std::max(i,j);
+                edge_vec.push_back(std::make_tuple(min_node, max_node));
+            }
```

Comment:
```suggestion
        for (long row_idx = 0; row_idx < n_rows; row_idx++) {
            if (assignments[row_idx] == within_label) {
                long i, j;
                if (self) {
                    i = calc_row_idx(row_idx, n_samples) + int_offset;
                    j = calc_col_idx(row_idx, i, n_samples) + int_offset;
                } else {
                    i = row_idx % num_ref + int_offset;
                    j = static_cast<size_t>(row_idx / (float)num_ref + 0.001f) + num_ref + int_offset;
                }
                if (i > j) {
                    std::swap(i, j);
                }
                edge_vec.push_back(std::make_tuple(i, j));
            }
        }
```

Reply:
may need `#include <algorithm>` added to top of file

Reply:
Implemented in modified form in 4d6c1a2.